### ISSUE TITLE
Test against official DCPU16 emulator

### ISSUE THREAD

**Issue by ghost** [llvm-dcpu16/llvm-dcpu16#129]
It seems that notch has release some sort of binary of the official core.

http://0x10cforum.com/forum/m/4932880/viewthread/2822200-notch-applet-how-be-used

We should look into it and see if there is a way to test our generated code against it.

**Comment by ghost** [comment]
Official jar:
http://0x10c.com/highnerd/

A launcher script that can launch assembled binaries. Probably we should try to write a very simple helper libc that includes the very basic functions like malloc / putc / memcpy etc to let us test things.

**Comment by krasin** [comment]
Nice.

@netzapper is nos_linker output compatible with the emulator? It would be nice, if so. :)

**Comment by aubreyrjones** [comment]
I will find out tonight. It is already netzapper/nos_link#4.

If it _isn't_ compatible, I can promise you by the end of tonight, it will be. My girlfriend works tonight. :)

**Comment by ghost** [comment]
@krasin The "emulator" is just the CPU core, around which people must build an emulator. It has no debugging features right now, so ... we either need a simmple printf/putc function to output our debug information, or we need to wait. The nos_linker output I provided him seemed to work, but I don't have any way to validate that.

**Comment by krasin** [comment]
ok, I see

what is the recommended way to received official updates from Notch?

**Comment by ghost** [comment]
For the emulator or the specs? Notch's "support" is really lazy in that matter ... he only releases stuff if he feels like it and has no time table.

**Comment by krasin** [comment]
Both. You tend to learn these things earlier than me and I still can't find an announcement of this official emulator in this twitter. So, I guess he has some other way to announce updates.

**Comment by aubreyrjones** [comment]
@arbaal During the day, I'm really a Java games hacker. You want me to try to wrap the emulator core in something useful?

I saw one very basic wrapper already on the forum. It seems simple, reading in word from disk and placing them in a buffer. I didn't try it, though, since I'm doing Actual Work at the moment.

A debugger might be very nice, if the cpu core has a step function.

**Comment by aubreyrjones** [comment]
Here's the launcher somebody wrote: http://www.0x10cforum.com/forum/m/4932880/viewthread/2822885-launcher-for-using-custom-programs-official-dcpu16-emulator

**Comment by ghost** [comment]
Ahh, I did forgot to include the link to the launcher _g_. Thanks netzapper. If you have time and fun to write such a wrapper, feel free to do so. But right now it's low priority and we should work on getting a SDK running till the end of the week.

**Comment by aubreyrjones** [comment]
I just tested. There is only video IO. So, I manged to make the screen flash colors. After much fiddling. I think I corrupted a character table or something.

The byte order in most emulators is wrong.

The byte order is native Java, big-endian. The patch for nos_link was trivial.

Do you want the old incorrect output on a switch? Or, shall we only support the reference byte order?

I vote for reference only. Community tools should rapidly adapt to the One True Way.

**Comment by ghost** [comment]
You're sure it's big endian? I think we would need to fix that in LLVM/Clang too. About the reference: Yeah, only the correct way please.

Also: It would be great if it's possible to have a facility to make regression tests with the emulator:
1. Give the emulator program x
2. Run programm x
3. Detect a stalling prgramm (JMP PC ...)
4. Return the content of A (default return value) to STD-OUT.

That way we could make automated regression tests for simple computation stuff. (But I don't know if it's possible)

**Comment by aubreyrjones** [comment]
It appears native java byte order. Big endian. link output with little
endian encoding fails. With big endian, puts characters on screen.

(we need to talk about strings, btw. Character 0x0078 paints an x, black on
black. Invisible. Color is required in encoding.)

I can write an awesome debugger from notch core. But it comes with java
dependency. On the other hand, if you're playing 0x10c, you have java.

And jruby is awesome. And fast. And easy to embed and distribute. I
actually prefer it to c ruby.

Only shitty thing is the java startup time.
On Apr 18, 2012 12:31 AM, "Nils Hasenbanck" <
reply@reply.github.com>
wrote:

**Comment by ghost** [comment]
Just for info: There was a chat yesterday, in which _notch (officialy) said that he will consider/implement the following proposal "soon":

https://github.com/0x10cStandardsCommittee/0x10c-Standards/blob/master/MISC/Open_Letter_Notch.txt

Chatlog:

http://vps.thomascomputerindustries.com/logs/freenode/0x10c-dev/2012-04-17

**Comment by krasin** [comment]
So, all issues but word addressing seem to be addressed. not bad.

**Comment by ghost** [comment]
No, not really: Arithmetic operations with carry are not added ... so we need to pay special attention to that.

**Comment by aubreyrjones** [comment]
The processor described in that new spec is far, far more flexible.  And sadly, still going to be so dog slow that preemptive multitasking still isn't going to happen.

I'm gonna detour from nos_link for the moment (it's working, just doesn't strip functions) and work on a debugger.

Anyway, @arbaal, I can totally do that for you.

I had planned not to build a regular emulator, but rather a debugger with memory watches and breakpoints. I am just going to wrap Notch's core, trying to modify it absolutely minimally. Mostly I have just cleaned up decompiler generated formatting and changed method visibility. But, I will have to expose ram (which I was right, is "char[] ram = char[0x10000];") and register state. It's easy and non-invasive, I think.

I should also be able to easily do hot loading, so you could run nos_db constantly as a server and have tests loaded to it. This helps mitigate the shitty java startup time.

So, adding a halt detector and returning register value would be very easy.

**Comment by aubreyrjones** [comment]
Man, programming Java feels so much like work...

**Comment by aubreyrjones** [comment]
Nope. Not gonna happen.

I'm switching to jruby for this. I can't bear hacking verbose, rigid Java all day and then programming more of it for "fun" in my free time.

On the plus side, this will mean if nos_link and nos_db are bundled together, nos_link can borrow the jruby implementation from nos_db. Only downside: I just profiled, and in that configuration, nos_link with CRuby compiles fib.s in 0.75s, while JRuby took 1.5s. And most of that is jvm spool up. In my testing at work, jruby is every bit as fast as cruby when it gets going... so, it should be fine in the debugger.

But still, it should be fine for most people. And people who care (like you and I who are building and rerunning tools 1000 times a day) can install cruby and flip a switch in the configuration.

Or, there may be a native C DCPU that we can consider trustworthy enough to replace the Notch core. I really don't want to detour so far off course that we're writing specs on a Certification Test Suite for DCPU16 Emulators.

Anyway, I'll at least make the binary distribution of this thing easy to use. You won't need java build tools to get it running. Just java.

**Comment by ghost** [comment]
Yeah, you can switch to jruby, since we can ship that with our SDK, since that's common practice in the java world.

**Comment by aubreyrjones** [comment]
There's a jar in netzapper/nos_db that works as a trivial highnerd wrapper.  You don't need java build tools, just Java runtime. Don't bother cloning the repo at this point, there's nothing interesting going on.
